Source/Diagnostics/WarpXOpenPMD.cpp

```diff
@@ -0,0 +1,479 @@
+#include "WarpXOpenPMD.H"
+#include "FieldIO.H"  // for getReversedVec
+
+WarpXOpenPMDPlot::WarpXOpenPMDPlot(bool oneFilePerTS,
+                       const std::string& openPMDFileType)
+  :m_Series(nullptr),
+   m_OneFilePerTS(oneFilePerTS),
+   m_OpenPMDFileType(openPMDFileType)
+{}
+
+WarpXOpenPMDPlot::~WarpXOpenPMDPlot()
+{
+  if (nullptr != m_Series) {
+    m_Series->flush();
+    delete m_Series;
+  }
+}
+
+
+//
+//
+//
+void WarpXOpenPMDPlot::GetFileName(std::string& filename, int ts)
+{
+  std::string dir = "diags/";
+  std::string prefix="/plot";
+
+  // Write openPMD format: only for level 0
+  filename = std::string(dir);
+  filename.append(m_OpenPMDFileType);
+
+  if (m_OneFilePerTS)
+    filename += amrex::Concatenate(prefix, ts);
+  else
+    filename.append(prefix);
+
+  filename += std::string(".") + m_OpenPMDFileType;
+}
+
+
+void WarpXOpenPMDPlot::SetOpenPMDType(const std::string& type)
+{
+  m_OpenPMDFileType = type;
+}
+
+void WarpXOpenPMDPlot::SetStep(int ts)
+{
+  if (ts < 0)
+    return;
+
+  m_CurrentStep =  ts;
+
+  std::string filename;
+  if (m_OneFilePerTS) {
+    GetFileName(filename, ts);
+    Init(filename, openPMD::AccessType::CREATE);
+  } else { // one file
+    if (nullptr == m_Series) {
+      GetFileName(filename, -1);
+      Init(filename, openPMD::AccessType::CREATE);
+    }
+  }
+}
+
+void
+WarpXOpenPMDPlot::Init(const std::string& filename,
+               openPMD::AccessType accessType)
+{
+  if (m_Series != nullptr) {
+    m_Series->flush();
+    delete m_Series;
+    m_Series = nullptr;
+  }
+
+  if (amrex::ParallelDescriptor::NProcs() > 1) {
+    m_Series = new openPMD::Series(filename,
```

Comment:
```suggestion
    m_Series = std::make_unique<openPMD::Series>(filename,
```

Reply:
sorry, I forgot that `make_unique` is C++17... just `m_Series = std::unique<openPMD::Series>( new openPMD::Series(filename, accessType, amrex::ParallelDescriptor::Communicator()) )` then;

Reply:
update: no problem, WarpX requires C++14 now. Let's use `make_unique` :)